src/math_utils.rs

```diff
@@ -182,6 +185,176 @@ pub fn sqrt(n: &Felt252) -> Felt252 {
+    }
+    let root_1 =
+        (n.modpow(&((t + 1_u32) >> 1), prime) * d.modpow(&(m >> 1), prime)).mod_floor(prime);
+    let root_2 = prime - &root_1;
+    if root_1 < root_2 {
+        root_1
+    } else {
+        root_2
+    }
+}
+
+/* Disclaimer: Some asumptions have been taken based on the functions that rely on this function, make sure these are true before calling this function individually
+Adpted from sympy implementation, asuming:
+    - p is an odd prime number
+    - a.mod_floor(p) == a
+Returns the Legendre symbol `(a / p)`.
+
+    For an integer ``a`` and an odd prime ``p``, the Legendre symbol is
+    defined as
+
+    .. math ::
+        \genfrac(){}{}{a}{p} = \begin{cases}
+             0 & \text{if } p \text{ divides } a\\
+             1 & \text{if } a \text{ is a quadratic residue modulo } p\\
+            -1 & \text{if } a \text{ is a quadratic nonresidue modulo } p
+        \end{cases}
+*/
+fn legendre_symbol(a: &BigUint, p: &BigUint) -> i8 {
+    if a.is_zero() {
+        return 0;
+    };
+    if is_quad_residue(a, p).unwrap_or_default() {
+        1
+    } else {
+        -1
+    }
+}
+
+/* Computed from:
+small_trailing = [0] * 256
+for j in range(1,8):
+    small_trailing[1<<j::1<<(j+1)] = [j] * (1<<(7-j))
+*/
+const SMALL_TRAILING: [u64; 256] = [
+    0, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    6, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    7, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    6, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
+];
+// Ported from sympy implementation
+fn trailing(n: BigUint) -> u64 {
+    let oxff = BigUint::from(0xff_u32);
+    let low_byte = &n & &oxff;
+    if !low_byte.is_zero() {
+        return SMALL_TRAILING[low_byte.to_usize().unwrap()];
+    }
+    let mut n = n;
+    let z = n.bits();
+    if n == BigUint::one() << z {
+        return z;
+    }
+    if z < 300 {
+        let mut t = 8;
+        n >>= 8;
+        while (&n & &oxff).is_zero() {
+            n = &n >> 8;
+            t += 8;
+        }
+        return t + SMALL_TRAILING[(n & &oxff).to_usize().unwrap()];
+    }
+    let mut t = 0;
+    let mut p = 8_u64;
+    while (&n & BigUint::one()).is_zero() {
+        while (&n & ((BigUint::one() << p) - 1_u32)).is_zero() {
+            n >>= p;
+            t += p;
+            p *= 2;
+        }
+        p = num_integer::Integer::div_floor(&p, &2);
+    }
+    t
+}
```

Comment:
Let's just use the one in `BigUint`: https://docs.rs/num-bigint/0.4.3/num_bigint/struct.BigUint.html#method.trailing_zeros

Reply:
Alternatively, I think it would be safe, simpler and faster to go with this:
```suggestion
/* Computed from:
small_trailing = [0] * 256
for j in range(1,8):
    small_trailing[1<<j::1<<(j+1)] = [j] * (1<<(7-j))
*/
const SMALL_TRAILING: [u64; 256] = [
    0, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    6, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    7, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    6, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
    5, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0, 4, 0, 1, 0, 2, 0, 1, 0, 3, 0, 1, 0, 2, 0, 1, 0,
];
// Ported from sympy implementation
fn trailing(n: BigUint) -> u64 {
    let mut n_bytes_iter = n.to_bytes_le().iter();
    let zeros = n_bytes_iter.take_while(|x| x == 0).count().unwrap_or(0) as u64;
    let Some(non_zero) = n_bytes_iter.next() else {
        return zeros;
    };
    zeros + SMALL_TRAILING[non_zero];
}
```